Sources/PackageRegistry/RegistryClient.swift

```diff
@@ -655,50 +705,160 @@ public final class RegistryClient: Cancellable {
+                        } else {
+                            throw RegistryError.invalidResponseStatus(expected: [200], actual: response.statusCode)
+                        }
+                    }
                 }.mapError {
                     RegistryError.failedRetrievingRegistryPublishRequirements($0)
                 }
             )
         }
     }
 
+    public func publish(
+            registryURL: URL,
+            packageIdentity: PackageIdentity,
+            packageVersion: Version,
+            packageArchive: AbsolutePath,
+            packageMetadata: AbsolutePath?,
+            timeout: DispatchTimeInterval? = .none,
+            fileSystem: FileSystem,
+            observabilityScope: ObservabilityScope,
+            callbackQueue: DispatchQueue,
+            completion: @escaping (Result<PublishResult, Error>) -> Void
+        ) {
+            let completion = self.makeAsync(completion, on: callbackQueue)
+
+            guard let scopeAndName = packageIdentity.scopeAndName else {
+                return completion(.failure(RegistryError.invalidPackage(packageIdentity)))
+            }
+            guard var components = URLComponents(url: registryURL, resolvingAgainstBaseURL: true) else {
+                return completion(.failure(RegistryError.invalidURL(registryURL)))
+            }
+            components.appendPathComponents(scopeAndName.scope.description)
+            components.appendPathComponents(scopeAndName.name.description)
+            components.appendPathComponents(packageVersion.description)
+
+            guard let url = components.url else {
+                return completion(.failure(RegistryError.invalidURL(registryURL)))
+            }
+
+            // TODO: don't load the entire file in memory
+            guard let packageArchiveContent: Data = try? fileSystem.readFileContents(packageArchive) else {
+                return completion(.failure(RegistryError.failedLoadingPackageArchive(packageArchive)))
+            }
+            var metadataContent: String? = .none
+            if let packageMetadata = packageMetadata {
+                do {
+                    metadataContent = try fileSystem.readFileContents(packageMetadata)
+                } catch {
+                    return completion(.failure(RegistryError.failedLoadingPackageMetadata(packageMetadata)))
+                }
+            }
+
+            // TODO: add generic support for upload requests in Basics
+            var body = Data()
+            let boundary = UUID().uuidString
+            
+            // archive field
+            body.append(contentsOf: """
+            --\(boundary)
+            Content-Disposition: form-data; name=\"source-archive\"
+            Content-Type: application/zip
+            Content-Transfer-Encoding: base64
+            Content-Length: \(packageArchiveContent.count)
+
+            \(packageArchiveContent.base64EncodedString())
```

Comment:
I don't know if we need `\r` at the end of each line:

```swift
        Content-Disposition: form-data; name="source-archive"\r
        Content-Type: application/zip\r
        Content-Transfer-Encoding: base64\r
        Content-Length: \(sourceArchive.count)\r
        \r
        \(sourceArchive.base64EncodedString())\r
```

I have it in my prototype. 🤔 We can leave it out for now, but if there is problem parsing the request then this might be a reason.